### ISSUE TITLE
Support subdomain changes without user login/logout

### ISSUE THREAD

**Issue by timabbott** [zulip/zulip-mobile#3142]
Occasionally, we end up renaming a subdomain on zulipchat.com (often because of the company changing names, being acquired, etc.).  Currently, this results in the mobile app starting to get this error block on the server:
```
        logging.warning("User %s (%s) attempted to access API on wrong subdomain (%s)" % (                                 
            user_profile.email, user_profile.realm.subdomain, get_subdomain(request)))                                     
        raise JsonableError(_("Account is not associated with this subdomain"))                                            
```

Empirically, this seems to result in the mobile app posting to the server several times a minute trying to do things, and repeatedly getting this error.  It's also of course a not-great user experience as well.

We should probably do one or more of the following:
* (1) Handle this error from `/register` and effectively log the user out in the mobile app (and maybe trigger some sort of notification to the user).  It might make sense to add a `code` to this error condition in the server logic to support doing so; this would be a quick change on the server side that's maybe worth doing in any case.
* (2) We could have the server send an event just before the subdomain changes indicating the new subdomain; the client could then update its data structures and continue working against the new subdomain.  While I believe mobile clients will start getting blocked for their next API request, any clients with an active long-poll could be migrated (which certainly doesn't cover all of them, but would cover some).  I don't like this solution because it's unreliable.
* (3) We could add a little `previous_subdomains` data structure in the backend, and when you get a `/register` request to the wrong one, give a special error that the client can parse.  This would enable the client to know what the new domain is and automatically migrate the user, similar to solution (2).  (For context, the order of operations is currently to first look up the user by API key, and then verify the domain, which is how we're able to do this logging, so we certainly have the ability to set this up).  This seems more robust than the last solution.

**Comment by borisyankov** [comment]
I am a fan of (3) It is basically an HTTP 301 for the API.
When we get this type of error on an API call we will change the 'realm' value.

We should investigate (1) also, as I suspect we are still not handling gracefully unknown but permanent errors. I guess we should see if we can determine the permanent-ness of an unknown error... 

One such error that we are not yet handling - Yago reported in India, he did invalidate his API key but the app did not realize it is no longer able to login (permanently). Logging the user out is not terribly inconvenient and is the right behavior.

**Comment by timabbott** [comment]
Yeah, I think any "API key invalid" error should probably log the user out assuming their API key was invalidated.  Let's move the API key invalidated version of this to a separate issue.

**Comment by timabbott** [comment]
OK.  For (1), I assume one of the folks who do both backend and mobile development will do the PR, since it should be pretty simple on the backend side and it's mostly about actually testing the thing.

**Comment by timabbott** [comment]
(I should add that this is a reasonably common situation.  A lot of organizations are importing data from HipChat/Slack/etc., and because a lot of the time folks will try out Zulip first with the reasonable subdomain for their organization, and then want to reuse the subdomain for a "new" organization with all their historical data imported).  We try to mitigate it by telling people they need to rename their organization when we do an import, but despite that we definitely have users whose app is spamming this error all day.

**Comment by borisyankov** [comment]
Good to know that this error is more frequent than what I would have guessed.

I mentioned the "API key invalid" earlier because I think for (1) it is basically the same behavior and we should check if we can treat it like the same case, that is if we do purely a logout.

I am imagining from a coding perspective it might be better to have the response be always an error code with some additional data where to redirect instead of it being included in a valid 'register' response.
